Get west and manifest revisions from a configuration file + support any revision format + misc.

ulfalizer opened the pull request:
```
Get west and manifest revisions from a configuration file

Add configuration file support via the standard Python configparser
module, which uses a Git-like .ini format. Use it to save the west and
manifest revisions specified when running 'west init', and use the saved
revisions when updating the west and manifest repositories.

Configuration file paths are based on Git, for consistency with the
Git-like commands. See the FILES section in git-config(1). Mac OS and
Windows paths were checked by Carles Cufi with
'git config --list --show-origin'.

Add some new shorthands to _expand_shorthands() in project.py to factor
out some logic while removing the hardcoding of the west and manifest
upstream branches. The qual-* versions of shorthands give the full path
to refs (refs/heads/foo, remotes/origin/bar, etc.)

Fixes: #63
Fixes: #67 
```

```
commands: project: Support any kind of revision format

Switch to a simpler and more flexible approach for dealing with upstream
revisions, that works with any revision format (SHA, branch, qualified
ref (e.g. refs/heads/foo), or (annotated) tag):

 1. Fetch (just) the upstream revision, exactly as specified.

 2. Dereference it to a commit via FETCH_HEAD, and point 'manifest-rev'
    at it.

    Using FETCH_HEAD makes it possible to apply ^{commit} peeling (see
    git-rev-parse(1)), which makes things work for e.g. annotated tags
    too (git tag -a ...).

A regression (could be an improvement, depending on how you see it) is
that just the upstream revision is fetched now. It'd be simple to do a
full fetch if we'd want that later though.
```

```
commands: project: Improve behavior for an aborted initial 'fetch'

The old _fetch() behavior was to check out a detached HEAD at
'manifest-rev' if the repository was initialized in the same call to
_fetch(). This meant that no detached HEAD was checked out if the
initial fetch was aborted and then resumed.

Instead, always check out a detached head at 'manifest-rev' if nothing
is checked out in the repository (which is almost guaranteed to mean
that we're still in the 'git init' state). That makes aborting and
resuming safe.

Nothing being checked out can be detected by looking at HEAD, which will
point to a non-existing ref.

Fixes: #78 
```

```
commands: project: Use remote name from manifest instead of 'origin'

To be consistent with repo (though I haven't actually checked the
behavior myself).

Fixes: #73
```

```
Add color.ui option for turning colors on/off

Based on the corresponding git option (see git-config(1)).

'true'/'false' are the only supported values at the moment, but they
work the same as in git (output is colorized if color.ui = true and the
output file is a terminal). If color.ui isn't specified, it defaults to
true.

color.ui is not written to west/config, since it makes more sense as a
global option than a per-repository option. It can still be disabled in
west/config though, though ~/.westconfig would make more sense.

I was thinking of toggling colors at a higher level with colorama
at first, but colorama.init() makes it a bit tricky to get right for
both Linux and Windows. Probably not too bad to make it explicit like
this.

Add a _reset_colors() helper to log.py as well to factor out a tricky
and easily typo'd print().
```

```
Make configuration values always available

This is nice in case something tries to read a configuration value
before the configuration files have been read. This could happen for the
logging functions for example, and for tests.

The configuration is empty before the configuration files have been
read, so defaults must be provided. They must be anyway though, for the
code to be robust in case configuration files are missing.
```

```
commands: project: Allow West/manifest remote to be changed

Store the name of the remote used for West and manifest updates in
west/config, like Git does it. That way the remote can be switched
(after adding a remote).

Another option would be to store a URL, but it's less flexible and would
add hairy remote management code to rewrite remote URLs. It'd also make
it impossible for users to manually manage remotes in Git, as their
settings would get overwritten.
```

```
commands: project: Check if projects are up-to-date when rebasing

_update() updates the special west/ and manifest/ repositories and has
code for checking if HEAD contains all the commits from a particular
revision (is up-to-date re. the revision). Move that code into a
separate _up_to_date_with() function, and also use it when rebasing
normal projects.

This gives a less spammy message when running 'pull' and 'rebase' on
already up-to-date projects:

  $ west rebase
  === zephyr (zephyr/) is up-to-date with manifest-rev

This commit also removes the special error message when self-updating
fails, returning to plain _git() with check=True. A separate context
hint mechanism will be added after this commit instead.

We could probably get rid of the Remote.url field. That information is
already available in the Git repository (we could fetch it from there if
we want to show it), and we should respect the repository settings in
case they have been changed by the user.
```

```
commands: project: Give context-specific errors for failed self-updates

Add a context manager ('with foo:' thingy) for setting a context for a
block of code. The context is just some text that gets added to the
error whenever a Git command fails.

Use the context manager to point out whenever a failing Git command is
related to self-updates. That could be a bit tricky to debug otherwise.

Example error:

  FATAL ERROR: Command 'git rebase 'FETCH_HEAD^{commit}'' failed for
  manifest (west/manifest/), while running automatic self-update. Please
  fix the state of the repository, or pass --no-update to 'west
  fetch/pull' to skip updating the manifest and West for the duration of
  the command.
```

```
commands: project: Refactor self-update a bit

Just have two _update_manifest() and _update_west() functions and get
rid of the flags. Bit easier to follow.

_update() was a bit overly generic as a name too.
```

```
commands: project: Add a 'west clone' command

This is just a shorthand for 'west fetch' + 'west checkout -b <branch>'.

The name of the branch is based on the revision from the manifest. For
revisions that are qualified refs (refs/heads/foo), the last component
(foo) is used as the name.

For SHA revisions, it doesn't make much much sense to name the branch
after the SHA, so 'work' is used instead.

The name of the created branch(es) can be overriden by passing
'-s <branch name>' to 'west clone'.
```

```
bootstrap: Print some information during initalization

Makes it a bit clearer what's going on, and helps people find the
configuration file.

Example output:

    Initializing in /home/ulf/westwest
    === Cloning west repository from https://github.com/zephyrproject-rtos/west, rev. master ===
    Cloning into '/home/ulf/westwest/west/west'...
    remote: Enumerating objects: 803, done.
    remote: Total 803 (delta 0), reused 0 (delta 0), pack-reused 803
    Receiving objects: 100% (803/803), 239.55 KiB | 371.00 KiB/s, done.
    Resolving deltas: 100% (428/428), done.
    === Cloning manifest repisitory from https://github.com/zephyrproject-rtos/manifest, rev. master ===
    Cloning into '/home/ulf/westwest/west/manifest'...
    remote: Enumerating objects: 6, done.
    remote: Counting objects: 100% (6/6), done.
    remote: Compressing objects: 100% (6/6), done.
    remote: Total 25 (delta 0), reused 2 (delta 0), pack-reused 19
    Unpacking objects: 100% (25/25), done.
    === Initial configuration written to /home/ulf/westwest/west/config ===
    === West initialized ===

Maybe some helpful hint could be added to the end too.
```

carlescufi commented:
To be consistent with how Git behaves:
* on macOS you should use `~/.westconfig` just like in Linux.
* on Windows you should use `<drive>\Users\<username>\.westconfig`

ulfalizer commented:
> To be consistent with how Git behaves:
> 
> * on macOS you should use `~/.westconfig` just like in Linux.
> * on Windows you should use `<drive>\Users\<username>\.westconfig`

Yeah, I wouldn't mind doing that. I was going off https://github.com/zephyrproject-rtos/west/issues/67, but maybe it'd be better to just stick to the Git conventions.

@mbolivar 
Thoughts?

mbolivar commented:
> @mbolivar
> Thoughts?

As discussed on IRC, consistency with git is a go.

carlescufi commented:
@mbolivar other than the file locations, are you happy with the changes?

ulfalizer commented:
Pushed a new version that uses more Git-like paths.

I followed the `FILES` section in `git-config(1)`. Looks like there might be some custom stuff going on for the system-wide configuration on Windows though. I just put on what Carles saw on his machine.

carlescufi commented on src/west/main.py:
```diff
@@ -199,6 +199,9 @@ def main(argv=None):
         argv = sys.argv[1:]
     args, unknown = parse_args(argv)
 
+    # Read the configuration files
```
Should we try to create one for those using the old bootstrap?

ulfalizer replied:
You mean people who have already bootstrapped?

Defaults are used if the configuration file doesn't exist, so shouldn't need it.

ulfalizer commented on src/west/config.py:
```diff
@@ -0,0 +1,55 @@
+
+    # User-specific and repository-specific
+
+    files.append(os.path.expanduser('~/.westconfig'))
```
Yep, works on Windows too: https://docs.python.org/3/library/os.path.html#os.path.expanduser

The Python docs are pretty awesome. :)

carlescufi commented:
This looks fine to me as-is now. That said, you introduced some additional changes that were unrelated to the config file in the same commit from what I see. Would be good if you could keep those in separate commits in upcoming PRs.

carlescufi commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+    # on all systems. They're listed in git-config(1).
+    files.append(os.path.expandvars('$XDG_CONFIG_HOME/git/config')
+                 if 'XDG_CONFIG_HOME' in os.environ else
+                 os.path.expanduser('~/.config/git/config'))
```
s/git/west

ulfalizer replied:
Heh... whoops :)

Fixed.

ulfalizer commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+    if platform.system() == 'Linux':
+        files = ['/etc/westconfig']
+    elif platform.system() == 'Darwin':  # Mac OS
+        # This was seen on Carles' machine ($(prefix) = /usr/local)
```
Done

ulfalizer commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+        # This was seen on Carles' machine ($(prefix) = /usr/local)
+        files = ['/usr/local/etc/westconfig']
+    elif platform.system() == 'Windows':
+        # Seen on Carles' machine. Not sure how it got derived though.
```
Done

mbolivar commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+        # This was seen on a local machine ($(prefix) = /usr/local)
+        files = ['/usr/local/etc/westconfig']
+    elif platform.system() == 'Windows':
+        # Seen on a local machine. Not sure how it got derived though.
```
What does "Not sure how it got derived though." mean?

ulfalizer replied:
Derived by Git. See the `FILES` section in `git-config(1)`.

I'll remove it.

mbolivar commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+        files = [os.path.expandvars('%PROGRAMDATA%\\west\\config')]
+
+    # Probably doesn't hurt to check $XDG_CONFIG_HOME and ~/.config/west/config
+    # on all systems. They're listed in git-config(1).
```
I really don't like that but I don't think it should block the merge. My guess is Git works that way because it's a Linux program that happens to run on other platforms. The XDG_CONFIG_HOME and friends variables are part of the XDG base directory specification (https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) and should only apply on those platforms in my opinion.

It will also make the documentation more awkward (and surprising IMO) for macOS and Windows users.

carlescufi replied:
We can live this out of the Windows/macOS doc in my opinion.

ulfalizer replied:
I couldn't see anything about it being Linux-specific in that doc, but I might've missed it. If someone wants to set it on e.g. Mac OS, then more power to them, IMO.

I suspect the reason they do it is because it's nice, and take [Linus' stance on standards](https://lkml.org/lkml/2018/6/5/769). :D

mbolivar replied:
It's not nice, and neither that email from Linus

carlescufi replied:
I suspect Linus wouldn't write  that email today after his change of stance regarding several topics recently.

mbolivar commented on src/west/config.py:
```diff
@@ -0,0 +1,56 @@
+    global config
+
+    #
+    # Gather (potential) configuration file paths, roughly following Git
```
The reason that I specifically chose to put the configuration file locations in a module level docstring in issue https://github.com/zephyrproject-rtos/west/issues/67 is because we are going to have to document this behavior for users, and it opens the path to letting a sphinx pydoc directive help us do that in a clean way. Doing it in plain comments will mean we have to duplicate the information in .rst, which encourages docs getting stale.

carlescufi replied:
Agree with that, let's mirror your text from #67 with the new locations in this PR

ulfalizer replied:
Ah, didn't notice those little `'''` around it.

Yeah, nice to have it in the docstring. I'll add it.

ulfalizer replied:
Done

mbolivar replied:
> Ah, didn't notice those little ''' around it.

I guess you also missed the text "Here is a proposed top-level docstring" which appeared before it.

Thanks for making this change.

ulfalizer commented on src/west/_bootstrap/main.py:
```diff
@@ -195,6 +196,14 @@ def init_bootstrap(directory, args):
+    config = configparser.ConfigParser()
+    config['west'] = {'revision': args.west_rev}
+    config['manifest'] = {'revision': args.manifest_rev}
+    with open(os.path.join(directory, WEST_DIR, 'config'), 'w') as f:
```
Isn't that how Git behaves? It's the only thing that makes sense to me.

What should it do instead? Ignore the command-line parameters if there are system-wide settings?

carlescufi replied:
Git overwrites the other configs. I tried this by adding:

```
[core]
filemode = false
```
in my `~/.gitconfig`. After cloning a new repo, I get
```
filemode = true
```

in `repo/.git/config`

carlescufi replied:
Then again this might not be the best [example](https://stackoverflow.com/a/30394903/337201)

ulfalizer replied:
Yeah, repository settings override system-wide settings in Git. The system-wide settings are more like defaults.

It'd get messy if they didn't, because there'd be no way to override a setting that happens to be in a system-wide configuration file.

ulfalizer replied:
What you could do would be to skip adding those entries to `west/conf` if no revisions are passed to `west init`. Then system-wide defaults could kick in.

Comes down to what you think the user wants when they run `west init` without specifying the revision. Is it the defaults for the parameters, or is the system-wide defaults?

IMO, it's much less confusing it means the defaults for those parameters (otherwise, we'd have to change the documentation too). The only real "save" if it meant some system-wide default revision would be that people wouldn't have to give the revision they want explicitly, and it just doesn't seem worth it.

carlescufi replied:
@mbolivar so your expectation is that `west init` follows the url and revisions present in your global or system configuration file in the case where the user has not specified one in the command line?

mbolivar replied:
> @mbolivar so your expectation is that `west init` follows the url and revisions present in your global or system configuration file in the case where the user has not specified one in the command line?

That's my expectation, yes. I think if the user has explicitly bothered to set up a config file in a non-default location and tried to configure west or the manifest there, they clearly expect that to have some effect.

> IMO, it's much less confusing it means the defaults for those parameters (otherwise, we'd have to change the documentation too). The only real "save" if it meant some system-wide default revision would be that people wouldn't have to give the revision they want explicitly if it's different from master, and it just doesn't seem worth it.

I actually think it is more confusing to do what you have described, given that you've decided to put these settings in the west configuration file in this implementation. The config file behavior should be consistent, rather than splitting some config into "it's okay for these to be global" and other config into "this is local only".

If the west and manifest URLs don't fit this pattern, then I don't think they should come from the West config file system.

carlescufi replied:
>If the west and manifest URLs don't fit this pattern, then I don't think they should come from the West config file system.

I agree with that, and also I am puzzled by the fact that only the revision is there and not the URL.
But @mbolivar, what would you do in the case of command-line options present? I would personally assume that command-line options take the highest priority in this case, since the user is specifically calling west with a set of URLs and revisions.

mbolivar replied:
> But @mbolivar, what would you do in the case of command-line options present? I would personally assume that command-line options take the highest priority in this case, since the user is specifically calling west with a set of URLs and revisions.

I agree. In increasing order of precedence, I think the west and manifest URL and revision should be determined by:

1. Defaults hard-coded into the bootstrapper
2. Site-wide config files (if any)
3. User-wide config files
4. Existing wworkdir west config (this only applies in the "reinit" case, i.e. if ``west init`` is invoked in an existing wworkdir)
5. Explicit ``west init`` command line options

carlescufi replied:
>Comes down to what you think the user wants when they run west init without specifying the revision. Is it the defaults for the parameters, or the system-wide defaults?

I think the user wants the system-wide defaults there, otherwise he wouldn't have bothered to fill those in in the config file. Why not print it?

```
west init .
Cloning west from URL:<url> at revision:<revision>...
Cloning manifest from URL:<url> at revision:<revision>...
```

mbolivar replied:
> I think the west and manifest URL and revision should be determined by:

Just to clarify, this is all under the assumption that they're configurable via the west config file system.

There are other options as well, for example, we discussed using ``git config`` to set ``west.xxx`` variables in another PR.

carlescufi replied:
>There are other options as well, for example, we discussed using git config to set west.xxx variables in another PR.

I did not follow that discussion but it seems to me that if we are going to introduce a config file for west we might as well keep everything in it to avoid confusing users, who would then need to go search for options in both `~/.gitconfig` and `~/.westconfig`

mbolivar replied:
> I did not follow that discussion but it seems to me that if we are going to introduce a config file for west we might as well keep everything in it to avoid confusing users, who would then need to go search for options in both `~/.gitconfig` and `~/.westconfig`

I agree -- just pointing out that we did discuss alternative means to configure the remote URL and revision for these "special case" repositories in the past if it comes to it.

carlescufi replied:
This is not addressed yet by your patches right @ulfalizer?

ulfalizer replied:
Is putting a manifest revision in a global configuration file just so that people can type `west init` instead of `west init -mr <rev>` really something we need in practice? It feels very obscure and a bit weird to me, like putting a global clone URL into `git` so that people can type `git clone` instead of `git clone <url>`.

Changing the manifest revision via `west init` would already be possible. You'd just rewrite the configuration file.

I'm imagining needing to rerun `west init` as something that would be pretty rare by the way, like rerunning `git init` (can do that to reinitialize hooks and stuff). Maybe you have a different workflow in mind.

carlescufi replied:
>Is putting a manifest revision in a global configuration file just so that people can type west init instead of west init -mr <rev> really something we need in practice?

Maybe I'm missing why we need a config file at all then? if we do have it, then I think we should allow people overriding the defaults. If we don't, then we should probably store these elsewhere?

carlescufi replied:
If we don't allow changing west's and the manifest's revision and url then we should do like git:

```
[remote "origin"]
        url = https://github.com/zephyrproject-rtos/west
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
```

Those cannot be overridden by global config options I assume, so those are "local variables only"?

carlescufi replied:
So I would actually argue that we do it (again) just like Git does. Having some local-only options that are not read from the system or user config files because:

* I think Ulf has a point, I don't see why a revision or a URL should be permanently set in a configuration file, thinking about it again for the nth time. This feels like hardcoding the URL you pass to `git clone` or the branch that you pass to `git checkout`.
* Git does this by creating the config options above, we can do something similar so it is clear to the user that those are local-only
* We need to document the global options like Git does here: https://git-scm.com/docs/git-config so that the user knows what to expect

So in my view, I would approve this PR if:

1) We save both revision and URL in `west/config`
2) We document the Git options (maybe add a simple one like no colors)
3) We allow `west init` to be re-run after having changed the config file or passing new command-line params to it

ulfalizer replied:
> * We save both revision and URL in `west/config`
> * We document the Git options (maybe add a simple one like no colors)
> * We allow `west init` to be re-run after having changed the config file or passing new command-line params to it

Sounds good to me.

Re-running `west init` seems alright to me, for the (I imagine, at least) rare cases where you'd need it. It'd just be a shorthand for changing the configuration.

Maybe we could add a `git config`-like at some point instead. I don't think it's horrible to have users directly edit configuration files in the meantime at least, regardless of the shorthand, as long as they're documented.

mbolivar replied:
> * I think Ulf has a point, I don't see why a revision or a URL should be permanently set in a configuration file, thinking about it again for the nth time. This feels like hardcoding the URL you pass to `git clone` or the branch that you pass to `git checkout`.

We do hardcode the git clone address of both west and the manifest, though, is the thing.

So there already is a default, and I don't see why the user should be prevented from doing a site- or user-wide override to `https://my-git-server.mycompany.com/{west,manifest}` or `https://github.com/myusername/{west,manifest}`.

> * Git does this by creating the config options above, we can do something similar so it is clear to the user that those are local-only

That said, I don't think this is worth blocking the merge over and I'm OK putting them into west/config if it's clear from the syntax that they are local to the wworkdir.

> We allow west init to be re-run after having changed the config file or passing new command-line params to it

Sounds like we're agreed then: re-running ``west init`` in an existing wworkdir (what I've been calling "reinit", not a separate command by that name) should be supported and allow changing the config so subsequent ``west fetch`` etc. reflect the new values.



> Maybe we could add a `git config`-like at some point instead. I don't think it's horrible to have users directly edit configuration files in the meantime at least, regardless of the shorthand, as long as they're documented.

I think editing config files by hand is perfectly natural FWIW. I actually prefer that for git config sometimes.

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -754,15 +753,40 @@ def _expand_shorthands(project, s):
+                     'refs/heads/' + project.revision) \
+            .replace('(qual-remote-rev)',
+                     project.revision if _is_sha(project.revision) else \
+                     'remotes/origin/' + project.revision) \
```
Yup, hardcoded to `origin` because it hasn't been made configurable yet.

carlescufi replied:
Note: this will not be "configurable" but rather: "equal to the `remote` field in the manifest" as per https://github.com/zephyrproject-rtos/west/issues/73#issuecomment-429002132

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -754,15 +753,40 @@ def _expand_shorthands(project, s):
             .replace('(path)', project.path) \
             .replace('(abspath)', project.abspath) \
-            .replace('(revision)', project.revision) \
+            .replace('(qual-rev)',
```
Yeah, qualified refs would be nice to support.

Hmm... tags get a bit messy. Maybe the "refs/heads/..." hurts more than it helps here. Pretty sure it's always first in the lookup order anyway. I'll look into it.

carlescufi replied:
@ulfalizer you had a proposal to fix this you mentioned on IRC?

mbolivar replied:
> Hmm... tags get a bit messy. Maybe the "refs/heads/..." hurts more than it helps here. Pretty sure it's always first in the lookup order anyway. I'll look into it.

We really need tags to work properly. As discussed in the comments in https://github.com/zephyrproject-rtos/west/issues/39, a manifest which has all its revisions set to tags is how we are going to manage Zephyr releases in the multi-repo world.

carlescufi commented on src/west/_bootstrap/main.py:
```diff
@@ -195,6 +196,14 @@ def init_bootstrap(directory, args):
+    # Create an initial configuration file
+
+    config = configparser.ConfigParser()
+    config['west'] = {'revision': args.west_rev}
```
Why are you storing only `revision` but not `url` (or `remote`)?

ulfalizer replied:
Not using them anywhere yet. Will add them once a need comes up. Maybe we can just get the information from Git.

carlescufi replied:
OK so there's something I don't understand here. Why make `revision` configurable then but not `url`? is it because you can change the revision of a manifest/west installation withou the need for a "reinit"?

ulfalizer replied:
Because none of the git commands need to look at the URL yet. It'd be dead code.

If they ever do, then we add it then. :)

ulfalizer replied:
Maybe the reinit thing would use it. To be honest, I'm not even sure how it's supposed to work.

Are people expected to regularly rerun the bootstrapper to do stuff?

carlescufi replied:
OK, but if you implement the parsing of the config as suggested, then it would make sense to add the url so that it takes over if the user doesn't specify it via cmd-line options?

carlescufi replied:
>Are people expected to regularly rerun the bootstrapper to do stuff?

I think only `reinit` if we choose to implement that as per the current proposal.

carlescufi replied:
I believe that if we implement the feature according to the comments https://github.com/zephyrproject-rtos/west/pull/75#discussion_r225202907 and https://github.com/zephyrproject-rtos/west/pull/75#discussion_r225202122, then you would use:

* `revision` always, as you are already doing now (plus of course parsing the conf files)
* `url`/`remote`: when running `west init` only for now (until we decide what to do with `reinit`).

ulfalizer replied:
Didn't realize we had a `reinit` too. Thought it was rolled into `init`. Same thing goes for that one though.

Configuration-handling-wise, what Git does, and what I think makes the most sense, is to just not write configuration options that aren't inherently repository-specific into `.git/config`. That way, you automatically get them from any global configuration files instead, and from defaults if they don't appear in any configuration files.

With that scheme, there's no need to read the parent configuration files if you're just updating the repository settings. It's pretty neat.

For example, you can make `author` a repository setting if you really want to by adding it to `git/config` (skipping the `--global` flag to `git config`), but `git clone` won't write it to `git/config` (or look at it at all).

So, if we want to make the revision a global setting unless specified (which I still think is a bit obscure and weird though), the neat Git-like way would be to just not write the revision into `west/config` unless explicitly specified.

ulfalizer commented:
@carlescufi @mbolivar 
I refactored the code refactored to support any revision format (tags, qualified refs, etc.) and added fixes for  #73 and #78.

ulfalizer commented:
Fixed a goof-up in `project.py`: The code was assuming that the `.ini` sections for the projects existed.

ulfalizer commented:
Added support for disabling colors in https://github.com/zephyrproject-rtos/west/pull/75/commits/a35570f7ae6372ee7b077b6eec227474046431ad.

ulfalizer commented:
Hmz... gotta fix up the tests too. They just reach in without doing any setup.

ulfalizer commented:
Thought of a nicer way to do it. Will take some shuffling around though. W8.

ulfalizer commented:
Pushed a separate refactoring commit instead: https://github.com/zephyrproject-rtos/west/pull/75/commits/fb4f02becd875730f51a77758e1869221170c0b5

Much nicer.

ulfalizer commented:
Added the possibility to change the remote.

One objection might be that I put a remote name instead of a URL. See the commit message for why I think a URL would be a worse idea, even if we're trying to help.

ulfalizer commented:
https://github.com/zephyrproject-rtos/west/pull/64 has been rebased onto this series now too.

I've also added a generic method for setting an error context (just some text that gets appended to `git` errors):

    with _error_context("doing stuff"):
        do the stuff

That's now used to print the hint re. self-update operations failing.

carlescufi commented on src/west/_bootstrap/main.py:
```diff
@@ -195,6 +196,23 @@ def init_bootstrap(directory, args):
+    config = configparser.ConfigParser()
+
+    config['west'] = {
+        'remote': 'origin',
```
This should be `remote.name`?

mbolivar commented:
Hi @ulfalizer; thanks a lot for the update!

I still need to test out that the error I was seeing (where west update didn't work properly if the manifest was on a non-default branch) is around, but I think this is looking really close personally.

There are two remaning blockers for me:

- We shouldn't be respecting XDG_CONFIG_HOME on non-freedesktop.org
  platforms (i.e. macOS and Windows). macOS isn't Linux, and Windows
  isn't even Unix. Linus acting like a jerk doesn't mean we should emulate him.

- Looks like we still need config keys that will make it clear to
  users that the west/manifest revisions are not configurable globally.

  As another potential starting point, it could be nice for the bootstrapper's init()
  to load existing config and emit a warning if it's ignoring
  global values by using its defaults. (This would mean tracking if
  the west and manifest parameters were specified by the user at the
  command line and silencing the warning in that case). This would at
  least eliminate user surprise and point the way forward (i.e. go in locally
  and update the values).

carlescufi commented:
>Looks like we still need config keys that will make it clear to
users that the west/manifest revisions are not configurable globally.

I agree, and discussed this with @ulfalizer yesterday. My suggestions:

```
[west.local]
remote = origin
revision = master
[manifest.local]
remote = origin
revision = master
```
or
```
[local.west]
remote = origin
revision = master
[local.manifest]
remote = origin
revision = master
```

ulfalizer commented:
@carlescufi @mbolivar 
> Looks like we still need config keys that will make it clear to
> users that the west/manifest revisions are not configurable globally.

The west/manifest revisions are configurable globally. Anything is. They work exactly the same as any other configuration key when read.

The system is this: If a key is not found in a repository-specific configuration file, then it is looked up in the user-specific configuration file(s), and finally in the system-wide configuration file.

If you delete e.g. the west revision from your repository-specific configuration file and put it in a system-wide configuration file, then that will work perfectly fine. At that point, you definitely don't want a `local.` prefix on it.

This will clarify how the system works:

Say storing the west/manifest revisions globally turns out to be useful (though it seems weird to me), and that you want users to be able to inherit them without hand-editing their configuration files. To do that, say you add a flag `--inherit-revisions` to the bootstrapper.

What would the `--inherit-revisions` flag do? It'd read the configuration files, wouldn't it? Nope! It still wouldn't read any configuration files at all. All it would do would be to not write the west/manifest revision into the repository-specific configuration file. That way, they get inherited instead.

> We shouldn't be respecting XDG_CONFIG_HOME on non-freedesktop.org
> platforms (i.e. macOS and Windows). macOS isn't Linux, and Windows
> isn't even Unix. Linus acting like a jerk doesn't mean we should emulate him.

What's the harm, in practical terms? If I set `XDG_CONFIG_HOME` on Mac OS and it gets ignored, then what did we win? All you've done is made it harder to share settings between operating systems, for something that's generic enough to work on all of them. That standard isn't meant only for "freedesktop platforms" either.

If you can look past the annoying assholishness, I've never seen someone be so consistently reasonable and down-to-earth as Linus, to be honest.

carlescufi commented:
> The west/manifest revisions are configurable globally. Anything is. They work exactly the same as any other configuration key when read.

They are for all intents and purposes, local. Because you populate them every time.

> 
> The system is this: If a key is not found in a repository-specific configuration file, then it is looked up in the user-specific configuration file(s), and finally in the system-wide configuration file.
> 
> If you delete e.g. the west revision from your repository-specific configuration file and put it in a system-wide configuration file, then that will work perfectly fine. At that point, you definitely don't want a `local.` prefix on it.

I don't think deleting lines is the way to configure anything.

mbolivar commented:
> > What would the --inherit-revisions flag do? It'd read the configuration files, wouldn't it? Nope! It still wouldn't read any configuration files at all. All it would do would be to not write the west/manifest revision into the repository-specific configuration file. That way, they come from the user-specific and system-wide configuration files instead.
> 
> You would need to load the config files, yes.
> 
> > The west/manifest revisions are configurable globally. Anything is. They work exactly the same as any other configuration key when read.
> 
> They are for all intents and purposes, local. Because you populate them every time.

This is the sticking point for me as well. As a user, if I had set up a global config and then run 'west init' without any options, I would expect west init to use those settings. The fact that it doesn't (and doesn't even warn me that it isn't) is the surprising behavior I feel strongly we should avoid.

If something can get stored in a west config file, it should be respected by west's configuration system in a uniform way. I don't get why that is a controversial opinion. It seems like another way to spell the hypothetical `--inherit-revisions` might be `--respect-my-configuration-files-if-they-exist`, which when you think about it, really ought to be the default, no?

ulfalizer commented:
> They are for all intents and purposes, local. Because you populate them every time.

They work exactly like any other configuration setting. If someone wants to put them in a user or system configuration file, then that's 100% supported.

Something like the `--inherit-revisions` could be added if a real need comes up where you'd want the revisions to come from global configuration files. It still seems very obscure to me, and for very obscure, just deleting the lines from `west/config` might be perfectly fine.

To clear up a potential misunderstanding: In the `--inherit-revisions` case, you actually would need to read configuration files too, but only so the bootstrapper knows what revision to use.

I think a better name for that flag would be `--use-global-config` or the like, if we'd ever need it. That could get the URL globally as well.

For what it's worth, it seems more awkward to me to require everyone to be on the same network so that they can get the same global configuration files than to just give then a short `west init ...` command that points to the manifest and west versions to use.

ulfalizer commented:
Added a `west clone` command, which is a shorthand for `west fetch` + `west checkout -b work`.

See the commit message for some discussion re. the branch name ("hey, why not master?"). Suggestions welcome for other things to call it (or methods for deriving the name, though I don't think we should make it too magic).

ulfalizer commented:
Hmm... then again, it might not be that bad to name the branch created by `west clone` the same as the (last component of the) revision, provided the revision points to a branch or a tag (and isn't e.g. a SHA).

One thing that gets slightly messy is if you change the revision to point somewhere else later.

What do you think? Can you think of other gotchas?

ulfalizer commented:
By the way, having a default for the West URL makes perfect sense to me, since nearly everyone would want the canonical version.

We could have a default for the manifest URL too, if we want to tie West hard to the Zephyr repository. The repo-like is currently bundled with a bunch of Zephyr-specific commands anyway.

So, basically, what we have now makes sense to me.

The part I don't get is getting manifest URL/revision from a global configuration file. When would you ever want that? If you're working on some internal project with your own manifest, then it's just a single `west init -u <URL>` command that needs to be documented. I think that helps demystify the whole manifest business to people too, since they can go and check it.

I might be missing something. @carlescufi also seemed to find this a bit weird.

ulfalizer commented:
Changed `west clone` to base the branch name on the revision in the manifest. Now you get a branch named `foo` if the manifest has `revision: foo`, etc.

For SHAs, you still get a generic 'work'. Wouldn't make much sense to create a branch with a 40-digit hexadecimal name.

carlescufi commented:
@ulfalizer 

I don't think they are unrelated questions, because it all comes down to whether those configuration options ever really have an impact on the behavior of west. 
As I see it, and I think @mbolivar has stated the same in the multiple comments you either

* Read the configuration in the bootstrapper so that the user or system-wide options take effect when you run `west init`
or
* Rename the options so that they are shown as being local-only (and yes, I know they are technically not, but practically they are)

Regardless, please print the url and revision when cloning/fetching so that the user sees where the data is coming from

tejlmand commented:
@ulfalizer 

I think a good discussion is already ongoing with regards to west configurations options.
Are there any particular reason that you suddenly add `west clone` to this PR ?

If more things are added to same PR, the discussion has a risk of getting more noisy than it already is.

Although I do think the `west clone` is a good feature, then I would still like to see it in a separate PR, as it has no relation to this PR.

ulfalizer commented:
@tejlmand 
I would've gone for `west clone` in a separate PR too, but this seems to have turned into the everything-before-the-deadline PR somehow.

`west clone` was added due to the detached HEAD complaints. I wonder if having some documentation for users to go off would help.

ulfalizer commented:
Oh, and nothing rebases cleanly anymore, since there's so much stuff in here, so that's the real reason. Getting some of this in earlier would've been nice. I'm not the kind of guy who abandons stuff once it's in.

ulfalizer commented:
Added some `print()`s to `west init` to tell people what's going on.

(That one probably does rebase cleanly. I'd just forgotten you can open separate PRs at this point. ;)

ulfalizer commented:
Heh... spoke too soon. It didn't rebase cleanly either.

tejlmand commented:
> Hence, I would propose we merge this, publish 0.3.0 and start collecting feedback from users and their workflow with west so we can actually make informed decisions that reflect the thinking of the majority of users. This includes as well designing the workflow that works best with west and the manifest, so that we know exactly how users work with the manifest, how they fork the repo and how they manage their separate forests (as in set of trees).

Agree with @carlescufi here.
Let's get this PR in, and start collecting feed-back.

ulfalizer commented:
> I don't see how adding `west clone` to this PR helps in this case.
> If it was put in separate PR, and the code changes, as I see them, are really self-contained, then a second PR can easily be rebased on top, when this PR gets approved.

Yeah, you can add lazy bastard to the list of excuses. That one might not have been horrible to be honest, though I haven't checked.

I've changed the XDG stuff to only be checked on Linux, to get things rolling.

ulfalizer commented on src/west/_bootstrap/main.py:
```diff
@@ -189,17 +191,40 @@ def init_bootstrap(directory, args):
           os.path.join(directory, WEST_DIR, WEST))
 
-    clone(args.manifest_url, args.manifest_rev,
+    clone('manifest repisitory', args.manifest_url, args.manifest_rev,
```
Fixed

ulfalizer commented:
@mbolivar 
Fixed.

Had goofed up in `Make configuration values always available`, loading configuration values into the wrong `ConfigParser`.
